### ISSUE TITLE
Cloud Hypervisor block I/O perf doesn't scale with the number of writers like QEMU does 

### ISSUE THREAD

**Issue by skinsbursky** [cloud-hypervisor/cloud-hypervisor#6520]
**Describe the bug**
Testing Virtio block sequential write bandwidth with sysbench from within a guest revealed that QEMU provides a substantially better performance scalability for multiple writers, while cloud hypervisor provided bandwidth stops scaling after at max 8+ writers.

The results for `sysbench fileio` are: 

|Threads|1|8|15|23|30|
|---------|-|-|--|--|--|
| Host| 1485.33 | 1606.92 | 1904.16 | 1882.80 | 1904.49 |
| QEMU guest| 802.77 | 1574.90 | 1888.99 | 2017.18 | 2101.60 |
| Cloud hypervisor guest|  814.14 | 1543.49 | 1579.80 | 1539.15 | 1470.92 |

It worth mentioning sysbench fileio does fsync every 100 writes by default, which explains relatively low BW for such device as `Dell Ent NVMe AGN RI U.2 1.92TB`.

**To Reproduce**

Install sysbench into the guest and run this loop:
`for t in 1 8 15 23 30; do sysbench fileio run --threads=${t} --file-test-mode=seqwr --file-num=4 --file-total-size=50mb | grep written | awk '{print $3}'; done`

Collect the results for both QEMU and Cloud Hypervisor KVM guests.

**Version**

Output of `cloud-hypervisor --version`:
`cloud-hypervisor v38.0-166-gfa7a000db`

Did you build from source, if so build command line (e.g. features):

**VM configuration**

What command line did you run (or JSON config data):

QEMU:
`sudo qemu-system-x86_64 -machine q35,accel=kvm,nvdimm=on -smp 1 -m 2G -kernel /usr/lib/modules/kernel-uvm/vmlinux -drive file=focal-server-cloudimg-amd64.raw,if=virtio -append root=/dev/vda1 console=ttyS0 -nographic`

Cloud hypervisor:
`sudo cloud-hypervisor --kernel hypervisor-fw --disk path=focal-server-cloudimg-amd64.raw --cpus boot=1 --memory size=2G --cmdline "root=/dev/vda1 console=hvc0"`

Guest OS version details:
```
root@cloud:~# uname -a
Linux cloud 5.4.0-167-generic #184-Ubuntu SMP Tue Oct 31 09:21:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

Host OS version details:
Custom

**Logs**

Output of `cloud-hypervisor -v` from either standard error or via `--log-file`:
```
# cat /tmp/ch.log | grep -i "block.rs\|device_manager.rs"
cloud-hypervisor: 2.901450ms: <vmm> INFO:vmm/src/device_manager.rs:2363 -- Creating virtio-block device: DiskConfig { path: Some("focal-server-cloudimg-amd64.raw"), readonly: false, direct: false, iommu: false, num_queues: 1, queue_size: 128, vhost_user: false, vhost_socket: None, rate_limit_group: None, rate_limiter_config: None, id: Some("_disk0"), disable_io_uring: false, disable_aio: false, pci_segment: 0, serial: None, queue_affinity: None }
cloud-hypervisor: 3.017143ms: <vmm> INFO:vmm/src/device_manager.rs:2453 -- Using asynchronous RAW disk file (io_uring)
cloud-hypervisor: 3.028045ms: <vmm> INFO:virtio-devices/src/block.rs:624 -- Disk topology: DiskTopology { logical_block_size: 512, physical_block_size: 512, minimum_io_size: 512, optimal_io_size: 0 }
cloud-hypervisor: 3.055277ms: <vmm> INFO:vmm/src/device_manager.rs:2775 -- Creating virtio-rng device: RngConfig { src: "/dev/urandom", iommu: false }
cloud-hypervisor: 39.191020ms: <vmm> INFO:virtio-devices/src/block.rs:709 -- Changing cache mode to writethrough
cloud-hypervisor: 2.625835s: <vmm> INFO:virtio-devices/src/block.rs:709 -- Changing cache mode to writeback
```
Linux kernel output:
N/A

**Comment by likebreath** [comment]
What's your host kernel version? Does it support io_uring? 

Also, you will want to use the same guest kernel for the measurement. In your measurement, QEMU is using direct kernel boot while Cloud Hypervisor is using the stock kernel from the guest image.

Some other general comments below.

To avoid noises from the kernel cache (both host and guest), you normally need to use "direct" I/O for the measurement, e.g. enabling `--disk direct=on` for Cloud Hypervisor and use `--direct=1` with `fio` (or something equivalent with `sysbench`). 

You also will need to match the number of worker threads with the number of vcpus and num_queues of virtio-blk to get a fair view of how does the performance scale with the number of workers.

I'd recommend you use `fio` as the measurement tool and create a dedicate virtio-blk device for the measurement.

**Comment by liuw** [comment]
Cloud Hypervisor's block device is really simple. In this test setup you have, there is one thread that supports the disk. What is does is very minimal -- take requests from the ring, submit request to io_uring, and when notified, give responses back to the guest. There is not lock in CH to content.

I have not done any deep analysis, but I notice something interesting.

Cloud Hypervisor hammers io_uring really hard. It causes io_uring to create dozens of worker threads in the kernel. QEMU, not so much, probably due to some internal capping mechanism.

This is not to say that's the culprit. To reach a concrete conclusion, we should take a look at `perf`'s output.

**Comment by skinsbursky** [comment]
> What's your host kernel version? Does it support io_uring?

It's a custom build, `5.15.159`-based:
```
$ zcat /proc/config.gz | grep URING
CONFIG_IO_URING=y
```
> Also, you will want to use the same guest kernel for the measurement. In your measurement, QEMU is using direct kernel boot while Cloud Hypervisor is using the stock kernel from the guest image.

Fair enough. 
Here are the results for the following guest launch command (apparently, they are lower than with the default kernel from the image):
`sudo cloud-hypervisor/target/release/cloud-hypervisor --kernel /usr/lib/modules/kernel-uvm/vmlinux --disk path=focal-server-cloudimg-amd64.raw,queue_size=16384 --cpus boot=1 --memory size=2G --cmdline root=/dev/vda1 console=hvc0`

|Threads|1|8|15|23|30|
|---------|-|-|--|--|--|
| Cloud hypervisor guest|  746.84 | 1400.34 | 1462.67 | 1385.69 | 1333.76 |

> To avoid noises from the kernel cache (both host and guest), you normally need to use "direct" I/O for the measurement, e.g. enabling --disk direct=on for Cloud Hypervisor and use --direct=1 with fio (or something equivalent with sysbench).

Well, indeed. Using `fio` and disabling caching in both host and guest (by running `fio` in `direct` mode) yields better results for cloud hypervisor than QEMU (~100 MiB/sec vs ~80 MiB/sec) in this single VP setup.
But it's not really the point, as it's obvious that caching, changing the number of queues and their depth can change the Virtio performance.

I guess the major difference here is that `sysbench` does periodic `fsync` by default (once per 100 requests) and this leads to this situation where cloud hypervisor is lagging behind.
It worth mentioning, that looks like this periodic syncing is used by some software apps, so `sysbench` tries to reproduce some real-life scenario by default.
Also, using disk image in direct I/O mode with `sysbench` (i.e. direct I/O in host and cached - in guest) yields better results for both QEMU and cloud hypervisor runs, but the latter one is still lagging behind.

So, to summarize, it looks like cloud hypervisor block write results lag behind QEMU in a particular scenario: cached writes with periodic syncing. And it doesn't matter if host caching is used or not.

**Comment by liuw** [comment]
I dug deeper into the io_uring code path in CH. I don't think it is a good idea to set the `ASYNC` flag. That forces the requests to go into IO threads. It is better to let the kernel decide what to do.

A liburing developer agrees with my observation: https://github.com/axboe/liburing/issues/349#issuecomment-841615536.

But removing that flag alone is not enough. I still see a lot of worker threads spawned.

**Comment by liuw** [comment]
I did a quick experiment. There is a way to cap the number of worker threads. Changing that doesn't make a huge difference.

**Comment by skinsbursky** [comment]
Perhaps Rust ` io_uring` crate is less efficient than `liburing`.

Here is another experiment with the same test but with disk, connected in `uncached` mode (`cache=directsync` for QEMU, `direct=on` for CH):

The results for `sysbench fileio` are: 

|Threads|1|8|15|23|30|
|---------|-|-|--|--|--|
| Host| 1485.33 | 1606.92 | 1904.16 | 1882.80 | 1904.49 |
| QEMU guest| 1200.60 | 2245.07 | 2247.53 | 2289.00 | 2349.77 |
| Cloud hypervisor guest|  1132.23 | 1612.68 | 1517.68 | 1447.54 | 1289.39 |

Conclusions:
1. Results get better for a single threaded write for both VMMs,
2. Cloud hypervisor is still lagging behind by 6%-45% (the more threads the bigger the regression is). I guess this is because cloud hypervisor backend scales worse/less than QEMU's, 
3. QEMU results become better than host ones. I guess this is because fsync now becomes a no-op and the VM can use most of its memory for caching.

It worth mentioning, that increasing the number of VM VPs to `2` with disk in uncached mode yields the following results (for CH the number of queues was increased to `2`): 

|Threads|1|8|15|23|30|
|---------|-|-|--|--|--|
| Host| 1485.33 | 1606.92 | 1904.16 | 1882.80 | 1904.49 |
| QEMU guest| 1220.10 | 2268.37 | 2273.88 | 2266.77 | 2312.04 |
| Cloud hypervisor guest|  1278.31 | 2039.15 | 2079.34 | 2064.28 | 2086.37 |

Conclusions:
1. QEMU results didn't change,
2. Cloud hypervisor results got better for multi-thread writes,
3. Cloud hypervisor results are now also better that the host ones,
4. Cloud hypervisor is still lagging behind QEMU, but now the regression is less than 10% across the board.

**Comment by liuw** [comment]
Okay, it looks like with sufficient tuning, CH is on par with QEMU.

I have been playing with some changes on my side. I think there is still come improvements we can do. It is too early for me to post anything though.

If we're satisfy with the current result, we can close this issue for now.
